### PR TITLE
feat: Add bucket metrics option

### PR DESCRIPTION
**Pull request by jcolfej**
## Description
Allow to add S3 bucket metrics.

## Motivation and Context
See https://github.com/terraform-aws-modules/terraform-aws-s3-bucket/issues/85

## Breaking Changes
No breaking change.

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -421,3 +421,19 @@ resource "aws_s3_bucket_ownership_controls" "this" {
     aws_s3_bucket.this
   ]
 }
+
+resource "aws_s3_bucket_metric" "this" {
+  for_each = var.create_bucket ? { for m in var.bucket_metrics : m.name => m } : tomap({})
```
Comment (bryantbiggs) [inline main.tf:426]:
```suggestion
  for_each = { for k,v in var.bucket_metrics : k => v if var.create_bucket}
```

you'll need to update below for `name = each.value.name`

File: variables.tf [hunk 2]
```diff
@@ -171,3 +171,9 @@ variable "object_ownership" {
   type        = string
   default     = "ObjectWriter"
 }
+
+variable "bucket_metrics" {
+  description = "(Optional) Specifies bucket metrics."
+  type        = any
+  default     = []
```
Comment (bryantbiggs) [inline variables.tf:178]:
```suggestion
  default     = {}
```

Comment (jcolfej) [reply variables.tf:178]: Done ;)

File: main.tf (hunk 1 above)
Comment (jcolfej) [reply main.tf]: Ok it's modified.
I just modified to use the name as index, so we don't have to destroy/recreate each metric if you override a metric on top of the list of metrics.

File: main.tf [hunk 3]
```diff
@@ -453,3 +453,19 @@ resource "aws_s3_bucket_ownership_controls" "this" {
     aws_s3_bucket.this
   ]
 }
+
+resource "aws_s3_bucket_metric" "this" {
+  for_each = { for k, v in var.bucket_metrics : v.name => v if var.create_bucket }
+
+  bucket = aws_s3_bucket.this[0].bucket
+  name   = each.value.name
```
Comment (antonbabenko) [inline main.tf:461]:
```suggestion
  name   = each.key
```

**Comment by xnick123** [comment]
Hi, could this be continued? it is a nice feature

**Comment by narayanan-sugumar** [comment]
Interested in this feature.

**Comment by jcolfej** [comment]
I had made all the modifications requested, but my work was not merged. 😢 
And today, unfortunately, I can't reopen the PR...

**Comment by jcolfej** [comment]
Thanks @bryantbiggs for the reopening.
I see there are conflicts, I'll take care of it tonight ;)

File: main.tf [hunk 4]
```diff
@@ -709,6 +709,22 @@ resource "aws_s3_bucket_ownership_controls" "this" {
...
 }
 
+resource "aws_s3_bucket_metric" "this" {
+  for_each = { for k, v in var.bucket_metrics : v.name => v if var.create_bucket }
+
+  bucket = aws_s3_bucket.this[0].bucket
+  name   = each.key
+
+  dynamic "filter" {
+    for_each = length(keys(lookup(each.value, "filter", {}))) == 0 ? [] : [lookup(each.value, "filter", {})]
```
Comment (bryantbiggs) [inline main.tf:719]:
```suggestion
    for_each = try([each.value.filter], [])
```

File: main.tf [hunk 5]
```diff
@@ -709,6 +709,22 @@ resource "aws_s3_bucket_ownership_controls" "this" {
...
+
+    content {
+      prefix = lookup(filter.value, "prefix", null)
```
Comment (bryantbiggs) [inline main.tf:722]:
```suggestion
      prefix = try(filter.value.prefix, null)
```

File: main.tf [hunk 6]
```diff
@@ -709,6 +709,22 @@ resource "aws_s3_bucket_ownership_controls" "this" {
...
+      tags   = lookup(filter.value, "tags", null)
```
Comment (bryantbiggs) [inline main.tf:723]:
```suggestion
      tags   = try(filter.value.tags, null)
```

File: variables.tf [hunk 7]
```diff
@@ -202,6 +202,12 @@ variable "object_ownership" {
   default     = "ObjectWriter"
 }
 
+variable "bucket_metrics" {
+  description = "(Optional) Specifies bucket metrics."
+  type        = any
```
Comment (bryantbiggs) [inline variables.tf:207]:
```suggestion
  type        = list(any)
```

File: variables.tf [hunk 8]
```diff
@@ -202,6 +202,12 @@ variable "object_ownership" {
...
+  default     = {}
```
Comment (bryantbiggs) [inline variables.tf:208]:
```suggestion
  default     = []
```

**Comment by jcolfej** [comment]
Thanks @bryantbiggs for the suggestions. 👍🏻 
Now that try(...) function exist, it greatly simplifies the reading of the code!

**Pull request closed without merging**